### PR TITLE
EVM fixes to burn base gas fee and priority fee to miners

### PR DESCRIPTION
**Pull request by sieniven**
## Summary

- Fixes to align mining block rewards with [EIP-1559](https://ethereum.org/en/developers/docs/gas/)
- Reward the miner with the priority gas fees / tips
- Base gas fees are burnt to burn address in consensus
- evm_finalize_block will return the burnt gas fees and priority gas fees of the minted eth block (forkguarded with changi-intermediate-height-4)

## Implications

- Storage
  - [ ] Database reindex required
  - [ ] Database reindex optional
  - [ ] Database reindex not required
  - [x] None

- Consensus
  - [x] Network upgrade required
  - [x] Includes backward compatible changes
  - [x] Includes consensus workarounds
  - [ ] Includes consensus refactors
  - [ ] None

### CODE REVIEW

File: lib/ain-evm/src/evm.rs [hunk 1]
```diff
@@ -31,7 +31,7 @@ pub struct EVMServices {
     pub storage: Arc<Storage>,
 }
 
-pub type FinalizedBlockInfo = ([u8; 32], Vec<String>, u64);
+pub type FinalizedBlockInfo = ([u8; 32], Vec<String>, u64, u64);
```
Comment (prasannavl) [inline lib/ain-evm/src/evm.rs:34]: Let's turn this into a struct

**Comment by prasannavl** [comment]
I believe it's time to finalize and just prep for a rollback. Let's start removing changi intermediates. Once we get a synced to the tip chain with the PRs planned today, let's remove the glues.

File: lib/ain-evm/src/evm.rs [hunk 2]
```diff
@@ -250,16 +251,20 @@ impl EVMServices {
         }
 
         if ain_cpp_imports::past_changi_intermediate_height_4_height() {
+            let total_burnt_fees = U256::from(total_gas_used) * base_fee;
+            let total_priority_fees = total_gas_fees - total_burnt_fees;
             Ok((
                 *block.header.hash().as_fixed_bytes(),
                 failed_transactions,
-                total_gas_fees.try_into().unwrap(),
+                total_burnt_fees.try_into().unwrap(),
```
Comment (Bushstar) [inline lib/ain-evm/src/evm.rs:259]: Let's return these as Sats. Currently this looks to  be WEI.

File: src/masternodes/validation.cpp [hunk 3]
```diff
@@ -2444,7 +2444,13 @@ static void ProcessEVMQueue(const CBlock &block, const CBlockIndex *pindex, CCus
         RevertFailedTransferDomainTxs(failedTransactions, block, chainparams.GetConsensus(), pindex->nHeight, cache);
     }
 
-    cache.AddBalance(minerAddress, {DCT_ID{}, static_cast<CAmount>(blockResult.miner_fee / CAMOUNT_TO_GWEI)});
+    if (pindex->nHeight >= chainparams.GetConsensus().ChangiIntermediateHeight4) {
+        cache.AddBalance(Params().GetConsensus().burnAddress, {DCT_ID{}, static_cast<CAmount>(blockResult.total_burnt_fees / CAMOUNT_TO_GWEI)});
+        cache.AddBalance(minerAddress, {DCT_ID{}, static_cast<CAmount>(blockResult.total_priority_fees / CAMOUNT_TO_GWEI)});
```
Comment (Bushstar) [inline src/masternodes/validation.cpp:2449]: If EVM returns Sats then we can remove CAMOUNT_TO_GWEI.

File: src/masternodes/validation.cpp [hunk 4]
```diff
@@ -2444,7 +2444,13 @@ static void ProcessEVMQueue(const CBlock &block, const CBlockIndex *pindex, CCus
...
+    }
+    else {
+        cache.AddBalance(minerAddress, {DCT_ID{}, static_cast<CAmount>(blockResult.total_burnt_fees / CAMOUNT_TO_GWEI)});
```
Comment (Bushstar) [inline src/masternodes/validation.cpp:2452]: Remove CAMOUNT_TO_GWEI.

File: src/miner.cpp [hunk 5]
```diff
@@ -281,7 +285,12 @@ std::unique_ptr<CBlockTemplate> BlockAssembler::CreateNewBlock(const CScript& sc
 
         const auto blockHash = std::vector<uint8_t>(blockResult.block_hash.begin(), blockResult.block_hash.end());
 
-        xvm = XVM{0,{0, uint256(blockHash), blockResult.miner_fee / CAMOUNT_TO_GWEI}};
+        if (nHeight >= consensus.ChangiIntermediateHeight4) {
+            xvm = XVM{0,{0, uint256(blockHash), blockResult.total_burnt_fees / CAMOUNT_TO_GWEI, blockResult.total_priority_fees / CAMOUNT_TO_GWEI}};
```
Comment (Bushstar) [inline src/miner.cpp:289]: Remove CAMOUNT_TO_GWEI

File: src/miner.cpp [hunk 6]
```diff
@@ -281,7 +285,12 @@ std::unique_ptr<CBlockTemplate> BlockAssembler::CreateNewBlock(const CScript& sc
...
+        }
+        else {
+            xvm_changi = XVMChangiIntermediate{0,{0, uint256(blockHash), blockResult.total_burnt_fees / CAMOUNT_TO_GWEI}};
```
Comment (Bushstar) [inline src/miner.cpp:292]: Remove CAMOUNT_TO_GWEI.

File: lib/ain-evm/src/evm.rs (hunk 2 above)
Comment (sieniven) [reply lib/ain-evm/src/evm.rs:259]: Thanks for this, I think the returned value is in GWEI. Will make the changes to return Sats, then we can nicely abstract gwei denomination to be only the EVM side of things.

File: lib/ain-evm/src/evm.rs (hunk 1 above)
Comment (sieniven) [reply lib/ain-evm/src/evm.rs]: Resolved in commit f03c76eeb2e

File: lib/ain-evm/src/evm.rs (hunk 2 above)
Comment (sieniven) [reply lib/ain-evm/src/evm.rs]: Resolved in ee66ff567

File: src/masternodes/validation.cpp (hunk 4 above)
Comment (sieniven) [reply src/masternodes/validation.cpp]: Resolved in commit ee66ff567

File: src/masternodes/validation.cpp (hunk 3 above)
Comment (sieniven) [reply src/masternodes/validation.cpp]: Resolved in commit ee66ff567

File: src/miner.cpp (hunk 5 above)
Comment (sieniven) [reply src/miner.cpp]: Resolved in commit ee66ff567

File: src/miner.cpp (hunk 6 above)
Comment (sieniven) [reply src/miner.cpp]: Resolved in commit ee66ff567

**Review by Bushstar — DISMISSED** [review]

**Review by Bushstar — DISMISSED** [review]

**Review by Bushstar — APPROVED** [review]

**Pull request merged**